multicluster/controllers/multicluster/clusterset_controller.go

```diff
@@ -60,3 +122,179 @@ func (r *ClusterSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+
+	if !wellKnownClusterClaimIDExist {
+		return fmt.Errorf("cluster claim ID not configured for the cluster")
+	}
+
+	configExists := false
+	if r.IsMcsLeader {
+		for _, leader := range clusterSet.Spec.Leaders {
+			if strings.Compare(clusterClaimID, leader.ClusterID) == 0 {
+				configExists = true
+			}
+		}
+		if !configExists {
+			return fmt.Errorf("cluster not defined as leader in cluster set")
+		}
+		// also validate the namespace is the same
+		if strings.Compare(clusterSet.Spec.Namespace, clusterSet.GetNamespace()) != 0 {
+			return fmt.Errorf("cluster set namespace " + clusterSet.Spec.Namespace + " is different from " +
+				clusterSet.GetNamespace())
+		}
+	} else {
+		for _, member := range clusterSet.Spec.Members {
+			if strings.Compare(clusterClaimID, member.ClusterID) == 0 {
+				configExists = true
+			}
+		}
+		if !configExists {
+			return fmt.Errorf("cluster not defined as member in cluster set")
+		}
+	}
+
+	r.clusterID = common.ClusterID(clusterClaimID)
+	r.clusterSetID = common.ClusterSetID(clusterSetClaimID)
+	r.clusterSetConfig = clusterSet.DeepCopy()
+
+	return nil
+}
+
+func (r *ClusterSetReconciler) createOrUpdateMultiClusterSetOnLeaderCluster(clusterSet *multiclusterv1alpha1.ClusterSet) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.inboundLeaderMgr == nil {
+		r.inboundLeaderMgr = internal.NewInboundLeaderClusterManager(r.Client, r.clusterID, clusterSet.GetNamespace(), r.Log)
+	}
+}
+
+func (r *ClusterSetReconciler) updateMultiClusterSetOnMemberCluster(clusterSet *multiclusterv1alpha1.ClusterSet) error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	if r.remoteClusterMgr == nil {
+		r.remoteClusterMgr = internal.NewRemoteClusterManager(r.clusterSetID, r.Log, r.clusterID)
+		go func() {
+			r.Log.Info("Starting remote cluster manager", "clusterSetID", r.clusterSetID)
+			var err error
+			err = r.remoteClusterMgr.Start()
+			if err != nil {
+				r.Log.Error(err, "Error starting remote remote cluster manager")
+			}
+			r.mutex.Lock()
+			r.remoteClusterMgr = nil
+			r.clusterSetID = common.INVALID_CLUSTER_SET_ID
+			r.clusterID = common.INVALID_CLUSTER_ID
+			r.clusterSetConfig = nil
+			r.mutex.Unlock()
+		}()
+	}
+
+	if r.remoteClusterMgr == nil {
+		return nil
+	}
+
+	currentLeaders := r.remoteClusterMgr.GetRemoteClusters()
+	newLeaders := clusterSet.Spec.Leaders
+
+	var addedLeaders []*multiclusterv1alpha1.MemberCluster
+	var removedLeaders map[common.ClusterID]internal.RemoteCluster
+
+	for _, leader := range newLeaders {
+		leaderID := common.ClusterID(leader.ClusterID)
+		_, found := currentLeaders[leaderID]
+		if !found {
+			addedLeaders = append(addedLeaders, leader.DeepCopy())
+		} else {
+			// In the end currentLeaders will only have removed leaders
+			delete(currentLeaders, leaderID)
+		}
+	}
+
+	ch := make(chan error)
+	var wg sync.WaitGroup
+	wg.Add(len(addedLeaders))
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
+
+	for _, addedLeader := range addedLeaders {
```

Comment:
does this mean a member cluster will talk to multiple leaders? how common area to reflect this? I didn't quite get this part, Could you add more information on PR to tell the whole scenario? thanks.

Reply:
Yes that is right, as we discussed offline, I also added some comments, hopefully that will help.